### PR TITLE
New model updates Jul 4

### PR DESCRIPTION
**Pull request by yqjing**
### Description
We intend to make some changed to the model so that it better suits our tasks and datasets. 

- early stop mechanism redesign
- rolling window {"train": (date1, date2), "test": (date3 , date4)} mechanism redesign, to account for the delayed parameter update, design in a way such that a parameter controls whether online learning will be done (exists parameter updates in the test set)
- loss function redesign, to mitigate the impact of overreaction of the model to some feature vectors, e.g. mse -> the L1 norm loss (huber loss)

### Checklist

- [x] early stop mechanism redesign
- [x] rolling window delayed parameter update redesign (online learning)
- [x] loss function redesign

### Goal

- Merge the branch `my_model` to the main `master`.

### CODE REVIEW

**Comment by yqjing** [comment]
### Task1: early stop mechanism redesign

1. Evaluate the fitness based on the mse in the validation dataset. Simplest solution: Stop based on a condition.  $mse^\*$ if 
$\mathcal{J}: mse < mse^\*$, break out the training and do inference in the test set. 

2. Based on the speed XGBoost fits to the data (within 5 epochs), and given the existence of delta feature, the model shall fit the data extremely fast (within 20 epochs probably). We let the model keep training until the condition is met. 

### Side-project

1. Store the checkpoint of the model that satisfies the condition.

**Comment by yqjing** [comment]
### Task2: rolling window parameter update mechanism redesign 

1. Switch from the "saving updated model (parameters)" to "natural way the model works" (still 20 days parameter update gap) (didn't work) (done)
4. adjust reg strength using sigma and reg given (done)
5. regularize the magnitude of delta_hat from one side, try maybe huber loss (done)

### Side-project:

1. get rid of the dependence on the saved prediction model and the related stuff, reference: 
   - project2_test_4/qlib/contrib/model/pytorch_alstm.py/class: ALSTM(Model)
   - project2_test_1/outputs/run_5-16002.out
   - get rid of the dependence of qlib

6. Plot the distribution of delta & error, to see if it is long tail. The loss thus needs to switch to the absolute loss.

**Review by yqjing — COMMENTED** [review]
Pls get away the qlib from my sight ASAP. 

Plus, solve the big problem of the parameter update mechanism in Task 2 pls.

**Comment by yqjing** [comment]
### Task3: Loss Redesign

1. Done, impose 1.5 times mse to case where y_hat > y_true, and 0.5 times mse to case where otherwise.

**Comment by yqjing** [comment]
I don't dare to **merge pull request** until I figure it out very clearly. I'm a bit worried the action will delete my **New model** branch which is the only backup I have for the forecasting model other than the one in the server (which might disappear someday somehow).  

I believe the best way to figure this out is to call XiaoZhang once since it's about the right time.
